### PR TITLE
Api 94: add search after method when getting a list of products in the API

### PR DESCRIPTION
**Pull request by ahocquard**
[//]: <> (<3 Thanks for taking the time to contribute! You're awesome! <3)

[//]: <> (If you've never contributed to this repository before, please read https://github.com/akeneo/pim-community-dev/blob/master/.github/CONTRIBUTING.md)

**Description (for Contributor and Core Developer)**

[//]: <> (What does this Pull Request do? reference the related issue?)

When you request a list of resources in the API, the classical method of pagination is with an offset.

```
select * from pim_catalog_product OFFSET 100 LIMIT 100;
```
 However, this pagination has some serious drawbacks : 
- it can return duplicate records in the request, because some products could be updated/added between two requests
- the page are returned more and more slowly due to mysql (late row look-up if you want more details ;) )

To avoid those drawbacks, another method to iterate on products has been implemented.
It has not been implemented on the other entities because the two drawbacks explained above does not concern those entities.

The `search after`method perform a request based on the last id returned in the previous request :
```
select * from pim_catalog_product where id > my_last_id
```

We do not expose directly the previous id in the url, because it is encrypted (approved by CTO).
We do not use the code because there was some serious performance problem with the fact that the code is in the PV.

Note : This method could not be used without ordering the products by id. 

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Ok
| Added Behats                      | -
| Added integration tests           | Ok
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/ApiBundle/Controller/AttributeController.php [hunk 1]
```diff
@@ -136,31 +144,76 @@ public function getAction(Request $request, $code)
...
         try {
-            $this->parameterValidator->validate($queryParameters);
+            $this->parameterValidator->validate($request->query->all());
         } catch (PaginationParametersException $e) {
             throw new UnprocessableEntityHttpException($e->getMessage(), $e);
         }
 
-        $offset = $queryParameters['limit'] * ($queryParameters['page'] - 1);
-        $attributes = $this->repository->searchAfterOffset([], [], $queryParameters['limit'], $offset);
+        if (PaginationTypes::OFFSET === $paginationType) {
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/AttributeController.php]: you should declare all commons variable before condition and create two dedicated methods for pagination to increase readability

File: src/Pim/Bundle/ApiBundle/Controller/AttributeController.php [hunk 2]
```diff
@@ -136,31 +144,76 @@ public function getAction(Request $request, $code)
...
+                ['code' => $queryParameters['search_after']]
+            );
+
+            $nbAttributes = count($attributes);
+
+            if ($nbAttributes === (int) $queryParameters['limit']) {
+                $lastAttribute = $attributes[$nbAttributes - 1];
+                $nextCode = $lastAttribute->getCode();
+            } else {
+                $nextEncryptedId = null;
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/AttributeController.php]: seems not used

File: src/Pim/Bundle/ApiBundle/Controller/AttributeController.php [hunk 3]
```diff
@@ -136,31 +144,76 @@ public function getAction(Request $request, $code)
...
+                [], ['code' => 'ASC'],
+                $queryParameters['limit'],
+                ['code' => $queryParameters['search_after']]
+            );
+
+            $nbAttributes = count($attributes);
+
+            if ($nbAttributes === (int) $queryParameters['limit']) {
+                $lastAttribute = $attributes[$nbAttributes - 1];
+                $nextCode = $lastAttribute->getCode();
```
Comment (skeleton) [inline src/Pim/Bundle/ApiBundle/Controller/AttributeController.php]: $nextCode is declared only on this condition (maybe it's just a mistake between $nextEncryptedId & $nextCode ?).
but, imvho, this code can be done in searchAfterPaginator service, it's his responsability to define if you have to have a next link
(FYI, in PHP, you can set `end($attributes)` to get the last element of an array)

File: src/Pim/Component/Api/Pagination/ParameterValidator.php [hunk 4]
```diff
@@ -29,38 +29,54 @@ public function __construct(array $configuration)
...
     /**
-     * @param int|string $limit
+     * @param array $parameters
      *
      * @throws PaginationParametersException
      */
-    protected function validateLimit($limit)
+    protected function validateLimit(array $parameters)
     {
+        if (!isset($parameters['limit'])) {
```
Comment (skeleton) [inline src/Pim/Component/Api/Pagination/ParameterValidator.php]: why ? we should check if a limit is set no ?

File: src/Pim/Component/Api/Pagination/SearchAfterHalPaginator.php [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    {
+        $this->resolver = new OptionsResolver();
+        $this->resolver->setDefaults([
+            'uri_parameters'      => [],
+            'item_identifier_key' => 'code',
+        ]);
+
+        $this->resolver->setRequired([
+            'query_parameters',
+            'search_after_identifiers',
```
Comment (skeleton) [inline src/Pim/Component/Api/Pagination/SearchAfterHalPaginator.php]: why not just "search_after" ?

File: src/Pim/Component/Api/Pagination/ParameterValidator.php (hunk 4 above)
Comment (ahocquard) [inline src/Pim/Component/Api/Pagination/ParameterValidator.php]: We validate the user input (and not the data before sending it to the repo), and limit is not mandatory in this case.

File: src/Pim/Bundle/ApiBundle/Controller/AttributeController.php (hunk 2 above)
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/AttributeController.php]: nice catch, not tested yet...

File: src/Pim/Bundle/ApiBundle/Controller/AttributeController.php (hunk 3 above)
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/AttributeController.php]: then , I will encrypt it for the products and add a parameter for that.

File: src/Pim/Bundle/ApiBundle/Controller/ProductController.php [hunk 6]
```diff
@@ -632,4 +635,90 @@ protected function validateCodeConsistency($code, array $data)
...
+     *
+     * @return array
+     */
+    protected function searchAfter(
+        ProductQueryBuilderInterface $pqb,
+        array $queryParameters,
+        array $normalizerOptions
+    ) {
+        $encryptedId = isset($queryParameters['search_after']) ? $queryParameters['search_after'] : null;
+        $id = null != $encryptedId ? $this->primaryKeyEncrypter->decrypt($encryptedId) : null;
```
Comment (phaseinducer) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: Why not `!==`?

File: src/Pim/Bundle/ApiBundle/Controller/ProductController.php [hunk 7]
```diff
@@ -632,4 +635,90 @@ protected function validateCodeConsistency($code, array $data)
...
+
+        $productCursor = $this->productRepository->searchAfterIdentifier($pqb, $queryParameters['limit'], $id);
+        $count = $this->productRepository->count($pqb);
+
+        $products = [];
+        foreach ($productCursor as $product) {
+            if (false !== $product) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: Can you explain please?

File: src/Pim/Bundle/ApiBundle/Doctrine/ORM/Repository/ProductRepository.php [hunk 8]
```diff
@@ -46,7 +46,11 @@ public function searchAfterOffset(ProductQueryBuilderInterface $pqb, $limit, $of
     {
         $qb = clone $pqb->getQueryBuilder();
 
+        $aliases = $qb->getRootAliases();
+        $rootAlias = $aliases[0];
```
Comment (phaseinducer) [inline src/Pim/Bundle/ApiBundle/Doctrine/ORM/Repository/ProductRepository.php]: `$rootAlias = $qb->getRootAliases()[0];` is ok ;)

File: src/Pim/Bundle/ApiBundle/tests/integration/Controller/Attribute/ListAttributeIntegration.php [hunk 9]
```diff
@@ -1,581 +1,107 @@
 <?php
 
-namespace Pim\Bundle\ApiBundle\tests\integration\Controller\Attribute;
+namespace Pim\Bundle\ApiBundle\tests\integration\Controller\Rest\Attribute;
```
Comment (phaseinducer) [inline src/Pim/Bundle/ApiBundle/tests/integration/Controller/Attribute/ListAttributeIntegration.php]: Sure?

File: src/Pim/Component/Api/Pagination/OffsetHalPaginator.php [hunk 10]
```diff
@@ -91,58 +97,38 @@ public function paginate(array $items, array $parameters, $count)
...
+     * Create the link to the self page.
      *
      * @param string $routeName
      * @param array  $parameters
-     * @param string $linkName
      *
      * @return Link
      */
-    protected function createLink($routeName, array $parameters, $linkName)
+    protected function createSelfLink($routeName, array $parameters)
```
Comment (phaseinducer) [inline src/Pim/Component/Api/Pagination/OffsetHalPaginator.php]: useless imho

File: src/Pim/Bundle/ApiBundle/Controller/ProductController.php (hunk 6 above)
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: no reason ;)

File: src/Pim/Bundle/ApiBundle/Controller/ProductController.php (hunk 7 above)
Comment (ahocquard) [inline src/Pim/Bundle/ApiBundle/Controller/ProductController.php]: I need to get the last element of the products and so, to iterate through the cursor. But with mongo, it's a collection. And this collection can have false due to mongo "bug". It's done like that in the normalizer too.

File: src/Pim/Component/Api/Pagination/ParameterValidator.php [hunk 11]
```diff
@@ -27,40 +27,60 @@ public function __construct(array $configuration)
...
     {
-        if (!isset($parameters['page'])) {
-            throw new PaginationParametersException('Page number is missing.');
+        if (!isset($parameters['pagination_type'])) {
+            $parameters['pagination_type'] = PaginationTypes::OFFSET;
         }
 
-        if (!isset($parameters['limit'])) {
-            throw new PaginationParametersException('Limit number is missing.');
+        switch ($parameters['pagination_type']) {
```
Comment (ahocquard) [inline src/Pim/Component/Api/Pagination/ParameterValidator.php]: And then the logic to handle default/valid pagination type would be done in every controllers... Much clearer like that.

File: src/Pim/Component/Api/Pagination/PaginationTypes.php [hunk 12]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace Pim\Component\Api\Pagination;
+
+use Pim\Component\Api\Exception\PaginationParametersException;
+use Pim\Component\Api\Hal\HalResource;
+use Pim\Component\Api\Hal\Link;
```
Comment (skeleton) [inline src/Pim/Component/Api/Pagination/PaginationTypes.php]: remove useless dependencies

File: src/Pim/Component/Api/Pagination/SearchAfterHalPaginator.php [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    public function paginate(array $items, array $parameters, $count)
+    {
+        try {
+            $parameters = $this->resolver->resolve($parameters);
+        } catch (\InvalidArgumentException $e) {
+            throw new PaginationParametersException($e->getMessage(), $e->getCode(), $e);
+        }
+
+        $data = [
+            'pages_count'  => $this->getLastPage($parameters['query_parameters']['limit'], $count),
```
Comment (skeleton) [inline src/Pim/Component/Api/Pagination/SearchAfterHalPaginator.php]: remove pages_count & items_count, it's useless on this king of pagination

Comment (skeleton) [inline src/Pim/Component/Api/Pagination/SearchAfterHalPaginator.php]: and remove method count() in controller too

**Pull request merged**